daemon/containerd/image.go

```diff
@@ -87,116 +90,56 @@ func (i *ImageService) getImageV1(ctx context.Context, img containerdimages.Imag
-	platform := matchAllWithPreference(platforms.Default())
+	pm := matchAllWithPreference(platforms.Default())
 	if options.Platform != nil {
-		platform = platforms.Only(*options.Platform)
+		pm = platforms.Only(*options.Platform)
 	}
 
-	cs := i.content
-
-	img, err := i.resolveImage(ctx, refOrID)
+	im, err := i.getBestPresentImageManifest(ctx, img, pm)
 	if err != nil {
 		return nil, err
 	}
 
-	desc := img.Target
-	if containerdimages.IsManifestType(desc.MediaType) {
-		plat := desc.Platform
-		if plat == nil {
-			config, err := img.Config(ctx, cs, platform)
-			if err != nil {
-				return nil, err
-			}
-			var configPlatform ocispec.Platform
-			if err := readConfig(ctx, cs, config, &configPlatform); err != nil {
-				return nil, err
-			}
-
-			plat = &configPlatform
-		}
-
-		if options.Platform != nil {
-			if plat == nil {
-				return nil, errdefs.NotFound(errors.Errorf("image with reference %s was found but does not match the specified platform: wanted %s, actual: nil", refOrID, platforms.Format(*options.Platform)))
-			} else if !platform.Match(*plat) {
-				return nil, errdefs.NotFound(errors.Errorf("image with reference %s was found but does not match the specified platform: wanted %s, actual: %s", refOrID, platforms.Format(*options.Platform), platforms.Format(*plat)))
-			}
-		}
+	desc := im.Target()
+	return &desc, nil
+}
 
-		return &desc, nil
-	}
+// getBestPresentImageManifest returns a platform-specific image manifest that best matches the provided platform matcher.
+// Only locally available platform images are considered.
+// If no image manifest matches the platform, an error is returned.
+func (i *ImageService) getBestPresentImageManifest(ctx context.Context, img containerdimages.Image, pm platforms.MatchComparer) (*ImageManifest, error) {
+	var best *ImageManifest
+	var bestPlatform ocispec.Platform
 
-	if containerdimages.IsIndexType(desc.MediaType) {
-		childManifests, err := containerdimages.LimitManifests(containerdimages.ChildrenHandler(cs), platform, 1)(ctx, desc)
+	err := i.walkImageManifests(ctx, img, func(im *ImageManifest) error {
+		imPlatform, err := im.ImagePlatform(ctx)
 		if err != nil {
-			if cerrdefs.IsNotFound(err) {
-				return nil, errdefs.NotFound(err)
-			}
-			return nil, errdefs.System(err)
+			return err
 		}
 
-		// len(childManifests) == 1 since we requested 1 and if none
-		// were found LimitManifests would have thrown an error
-		if !containerdimages.IsManifestType(childManifests[0].MediaType) {
-			return nil, errdefs.NotFound(fmt.Errorf("manifest has incorrect mediatype: %s", childManifests[0].MediaType))
+		if pm.Match(imPlatform) {
+			if best == nil || pm.Less(imPlatform, bestPlatform) {
+				best = im
+				bestPlatform = imPlatform
+			}
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
 
-		return &childManifests[0], nil
+	if best == nil {
+		return nil, &errPlatformNotFound{imageRef: imageFamiliarName(img)}
```

Comment:
should we not set the wanted platform here?

Reply:
Nevermind, saw that we're doing this in the caller. Wonder if we could move that.

Reply:
I see. I don't love having to remember to do that, but oh well 🤷‍♀️ Maybe we could mention that in `getBestPresentImageManifest`'s godoc?

Reply:
Hmm I got a funny idea, let me try it out 🙈

Reply:
So, I changed `matchRequestedOrDefault` to return a wrapper that also has the requested platform bundled in and use that to populate the error when it's created, so the caller doesn't need to do it.

WDYT?

Reply:
Looks good to me, @thaJeztah?